arch/lkl/kernel/syscalls.c

```diff
@@ -33,400 +32,136 @@ syscall_handler_t syscall_table[__NR_syscalls] = {
-	lkl_put_irq(data->irq, "syscall");
+	host_task_id++;
 
-	if (data->stop) {
-		complete(&data->stopped);
-	} else {
-		s->ret = 0;
-		lkl_ops->sem_up(data->completion);
-	}
+	snprintf((*task)->comm, sizeof((*task)->comm), "host%d", host_task_id);
 
 	return 0;
 }
 
-static unsigned int syscall_thread_data_key;
-
-static int syscall_thread_data_init(struct syscall_thread_data *data,
-				    void *completion)
+static void del_host_task(void *arg)
 {
-	data->mutex = lkl_ops->sem_alloc(1);
-	if (!data->mutex)
-		return -ENOMEM;
+	struct task_struct *task = (struct task_struct *)arg;
 
-	if (!completion)
-		data->completion = lkl_ops->sem_alloc(0);
-	else
-		data->completion = completion;
-	if (!data->completion) {
-		lkl_ops->sem_free(data->mutex);
-		data->mutex = NULL;
-		return -ENOMEM;
-	}
+	lkl_cpu_get();
 
-	return 0;
+	switch_to_host_task(task);
+	host_task_id--;
+	thread_set_sched_exit();
+	do_exit(0);
 }
 
-static long __lkl_syscall(struct syscall_thread_data *data, long no,
-			  long *params)
-{
-	struct syscall s;
-
-	s.no = no;
-	s.params = params;
-
-	lkl_ops->sem_down(data->mutex);
-	data->s = &s;
-	lkl_trigger_irq(data->irq);
-	lkl_ops->sem_down(data->completion);
-	lkl_ops->sem_up(data->mutex);
+static unsigned int task_key;
 
-	return s.ret;
-}
-
-static struct syscall_thread_data *__lkl_create_syscall_thread(void)
+long lkl_syscall(long no, long *params)
 {
-	struct syscall_thread_data *data;
-	long params[6], ret;
-
-	if (!lkl_ops->tls_set)
-		return ERR_PTR(-ENOTSUPP);
-
-	data = lkl_ops->mem_alloc(sizeof(*data));
-	if (!data)
-		return ERR_PTR(-ENOMEM);
-
-	memset(data, 0, sizeof(*data));
-
-	ret = syscall_thread_data_init(data, NULL);
-	if (ret < 0)
-		goto out_free;
-
-	ret = lkl_ops->tls_set(syscall_thread_data_key, data);
-	if (ret < 0)
-		goto out_free;
+	static int count;
+	long ret;
 
-	params[0] = (long)data;
-	ret = __lkl_syscall(&default_syscall_thread_data,
-			    __NR_create_syscall_thread, params);
+	ret = lkl_cpu_get();
 	if (ret < 0)
-		goto out_free;
-
-	lkl_ops->sem_down(data->completion);
-
-	return data;
-
-out_free:
-	lkl_ops->sem_free(data->completion);
-	lkl_ops->sem_free(data->mutex);
-	lkl_ops->mem_free(data);
-
-	return ERR_PTR(ret);
-}
-
-int lkl_create_syscall_thread(void)
-{
-	struct syscall_thread_data *data = __lkl_create_syscall_thread();
-
-	if (IS_ERR(data))
-		return PTR_ERR(data);
-	return 0;
-}
-
-static int kernel_stop_syscall_thread(struct syscall_thread_data *data)
-{
-	data->stop = true;
-	wake_up(&data->wqh);
-	wait_for_completion(&data->stopped);
-
-	return 0;
-}
-
-static int __lkl_stop_syscall_thread(struct syscall_thread_data *data,
-				     bool host)
-{
-	long ret, params[6];
-
-	if (host)
-		ret = __lkl_syscall(data, __NR_reboot, params);
-	else
-		ret = kernel_stop_syscall_thread(data);
-	if (ret)
 		return ret;
 
-	lkl_ops->sem_free(data->completion);
-	lkl_ops->sem_free(data->mutex);
-	lkl_ops->mem_free(data);
+	count++;
 
-	return 0;
-}
+	if (lkl_ops->tls_get) {
```

Comment:
I see you point. Before auto_syscall_threads, all syscall are serialized. Now, without TLS, I don't see it works and there is no mechanism forcing "all syscalls will be directed to host0". It can be idle context in the case above. Let me know if I misunderstand it.